cip/CIP2015-06-24-call-procedures.adoc

```diff
@@ -0,0 +1,250 @@
+
+If a procedure call fails to execute (i.e. it "throws an exception"), this error is propagated to the user in the same way as other runtime errors are propagated to the user by the implementation.
+
+If executing a procedure call causes any side effects (i.e. it "updates the graph"), all such changes should be executed
+before any results are returned to the user. An implementation may provide the user with a way to opt out of this
+behavior, however this must be done explicitly (e.g. via a configuration setting).
+
+==== Semantics of explicit argument passing
+
+In this scenario, arguments are provided explicitly as a sequence of expressions as required by the procedure's signature.
+It is an error if the number of provided arguments differs from the number of arguments required by the procedure signature.
+
+To call the procedure, all argument expressions are evaluated to argument values in order.
+It is an error if the argument values are incompatible with the argument types required by the procedure signature.
+
+==== Semantics of implicit argument passing
+
+In this scenario, arguments are providede implicitly via the parameters of the Cypher statement.
+
+To call the procedure, the argument values are obtained by using the parameter in scope with the same name as the procedure argument.
+If such a parameter does not exist, the argument value is taken to be `null`.
+It is an error, if the resulting argument values are incompatible with the argument types required by the procedure  signature.
+
+== What others do
+
+The stored procedures survey (separate google doc) is extremely comprehensive, examining how stored procedures are
+implemented and deployed as well as their API access mechanisms and usage.
+Products surveyed include PostgreSQL, MS SQL Server, Oracle, MySQL, MongoDB, Aerospike and Virtuoso.
+
+== Benefits to this proposal
+
+The benefits of having user-defined procedures is so that users would be able to implement algorithms and functionality
+which Cypher cannot either express or which cannot be executed efficiently by current Cypher implementations. Additionally, users may find procedures to be a useful mechanism to achieve good system design and code abstraction.
+
+== Audience Consultation
+
+CLG and interested internal parties.
+
+== Caveats to this proposal
+
+Stored procedures are a powerful extension mechanism.
+Their introduction opens up new ways of using Cypher which over time may lead to suboptimal useage patterns and hard to read queries.
+The introduction of stored procedures therefore carries a risk of influencing the long term evolution of the language in a negative way.
+
+== Appendix: Calling modes
+
+.Calling modes
+[frame="topbot",cols="3l,2a,2a,2a,1m", options="header,footer"]
+|===============================================================================
+|Statement template        |Mode      |Argument Passing|New Bindings|Cardinality
```

Comment:
To aid readability of this extremely useful table, please add a legend or something similar explaining how some of the columns are to be interpreted; in particular, the `Cardinality` column